File: android2_lesson04/src/java/java/com/example/ama/android2_lesson04/ui/viewer/PVRxJavaFragment.java
package com.example.ama.android2_lesson04.ui.viewer;

import android.os.Bundle;
import android.support.annotation.Nullable;

import com.example.ama.android2_lesson04.background.utils.NetworkUtils;
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment;
import com.example.ama.android2_lesson04.utils.ResourcesUtils;

import java.util.Arrays;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class PVRxJavaFragment extends BasePictureViewerFragment {
    private Disposable disposable;

    public static PVRxJavaFragment newInstance() {
        return new PVRxJavaFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        setRetainInstance(true);
        super.onCreate(savedInstanceState);
    }

    @Override
    protected void onLoadImagesClick() {
        if (disposable == null) {
            loadWithRx(ResourcesUtils.getUrlsArray());
        }
    }

    private void loadWithRx(String... urls) {
        disposable = Observable.fromIterable(Arrays.asList(urls))
                .map(NetworkUtils::loadBitmapFromUrl)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnComplete(this::onFinishLoading)
                .doOnError(throwable -> onFinishLoading())
                .subscribe(this::onUpdateLoading);
    }

    @Override
    protected void onFinishLoading() {
        super.onFinishLoading();
        dispose();
    }

    @Override
    public void onDestroy() {
        dispose();
        super.onDestroy();
    }

    private void dispose() {
        if (disposable != null) {
            disposable.dispose();
            disposable = null;
        }
    }
}

File: android2_lesson04/src/kotlin/java/com/example/ama/android2_lesson04/ui/viewer/PVRxJavaFragment.kt
package com.example.ama.android2_lesson04.ui.viewer

import android.os.Bundle
import com.example.ama.android2_lesson04.R
import com.example.ama.android2_lesson04.background.utils.NetworkUtils
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class PVRxJavaFragment : BasePictureViewerFragment() {
    var disposable: Disposable? = null

    companion object {
        fun newInstance() = PVRxJavaFragment()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        retainInstance = true
        super.onCreate(savedInstanceState)
    }

    override fun onDestroy() {
        dispose()
        super.onDestroy()
    }

    override fun onFinishLoading() {
        super.onFinishLoading()
        dispose()
    }

    override fun onLoadImagesClick() {
        if (disposable == null) {
            loadWithRx()
        }
    }

    private fun loadWithRx() {
        disposable = Observable.fromIterable(listOf(*resources.getStringArray(R.array.image_urls)))
                .map { it -> NetworkUtils.loadBitmapFromUrl(it)!! }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError { onFinishLoading() }
                .doOnComplete { onFinishLoading() }
                .subscribe { it -> onUpdateLoading(it) }
    }

    private fun dispose() {
        disposable?.dispose()
        disposable = null
    }
}